xdcr.go

```diff
@@ -0,0 +1,245 @@
+			}
+
+			originalCas, err := toDataStore.Get(docID, nil)
+			if err != nil && !toDataStore.IsError(err, sgbucket.KeyNotFoundError) {
+				warn("Skipping replicating doc %s, could not perform a kv op get doc in toBucket: %s", event.Key, err)
+				r.errorCount.Add(1)
+				return false
+			}
+
+			/* full LWW conflict resolution is not implemented in rosmar yet
+
+			CBS algorithm is:
+
+			if (command.CAS > document.CAS)
+			  command succeeds
+			else if (command.CAS == document.CAS)
+			  // Check the RevSeqno
+			  if (command.RevSeqno > document.RevSeqno)
+			    command succeeds
+			  else if (command.RevSeqno == document.RevSeqno)
+			    // Check the expiry time
+			    if (command.Expiry > document.Expiry)
+			      command succeeds
+			    else if (command.Expiry == document.Expiry)
+			      // Finally check flags
+			      if (command.Flags < document.Flags)
+			        command succeeds
+
+
+			command fails
+
+			In the current state of rosmar:
+
+			1. all CAS values are unique.
+			2. RevSeqno is not implemented
+			3. Expiry is implemented and could be compared except all CAS values are unique.
+			4. Flags are not implemented
+
+			*/
+
+			if event.Cas <= originalCas {
+				trace("Skipping replicating doc %s, cas %d <= %d", docID, event.Cas, originalCas)
+				return true
+			}
+
+			toCollection, ok := toDataStore.(*Collection)
+			if !ok {
+				warn("Datastore is not of type Collection, is of type %T", toDataStore)
+				r.errorCount.Add(1)
+			}
+
+			err = writeDoc(ctx, toCollection, originalCas, event)
+			if err != nil {
+				warn("Replicating doc %s, could not write doc: %s", event.Key, err)
+				r.errorCount.Add(1)
+				return false
+
+			}
+			r.docsWritten.Add(1)
+		}
+
+		return true
+
+	}
+	return r.fromBucket.StartDCPFeed(ctx, args, callback, nil)
+}
+
+// Stop terminates the replication.
+func (r *XDCR) Stop(_ context.Context) error {
+	close(r.terminator)
+	r.terminator = nil
+	return nil
+}
+
+// writeDoc writes a document to the target datastore. This will not return an error on a CAS mismatch, but will return error on other types of write.
+func writeDoc(ctx context.Context, collection *Collection, originalCas uint64, event sgbucket.FeedEvent) error {
+	if event.Opcode == sgbucket.FeedOpDeletion {
+		_, err := collection.Remove(string(event.Key), originalCas)
+		if !errors.Is(err, sgbucket.CasMismatchErr{}) {
+			return err
+		}
+		return nil
+	}
+
+	var xattrs []byte
+	var body []byte
+	if event.DataType&sgbucket.FeedDataTypeXattr != 0 {
+		var err error
+		var dcpXattrs []sgbucket.Xattr
+		body, dcpXattrs, err = sgbucket.DecodeValueWithXattrs(event.Value)
+		if err != nil {
+			return err
+		}
+
+		xattrs, err = xattrToBytes(dcpXattrs)
+		if err != nil {
+			return err
+		}
+
+	} else {
+
+		body = event.Value
+
+	}
+
+	err := collection.SetWithMeta(ctx, string(event.Key), originalCas, event.Cas, event.Expiry, xattrs, body, event.DataType)
+
+	if !collection.IsError(err, sgbucket.KeyNotFoundError) {
```

Comment:
Why do we want to ignore a KeyNotFound error here?

Reply:
Good catch.